### PR TITLE
EDLY-2014 - Add restriction on saving django_settings_override in site configuration

### PR DESCRIPTION
**Pull request by rehan99000**
**Description:**
Adds restrictions through clean method of site configuration to make sure only allowed settings are being overridden.

**JIRA:**
https://edlyio.atlassian.net/browse/EDLY-2014

image
image

### CODE REVIEW

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
I have added my feedback.

File: openedx/features/edly/utils.py [hunk 1]
```diff
@@ -313,3 +314,35 @@ def filter_courses_based_on_org(request, all_courses):
...
+
+def clean_django_settings_override(django_settings_overrides):
+    """
+    Enforce only allowed django settings to be overridden.
+    """
+    if not django_settings_overrides:
+        return
+
+    django_settings_overrides_keys = django_settings_overrides.keys()
+    disallowed_overrides = list(set(django_settings_overrides_keys) - set(settings.ALLOWED_DJANGO_SETTINGS_OVERRIDE))
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply openedx/features/edly/utils.py:326]: done

File: openedx/features/edly/utils.py [hunk 2]
```diff
@@ -313,3 +314,35 @@ def filter_courses_based_on_org(request, all_courses):
...
     return filtered_courses
+
+def clean_django_settings_override(django_settings_overrides):
+    """
+    Enforce only allowed django settings to be overridden.
+    """
+    if not django_settings_overrides:
+        return
+
+    django_settings_overrides_keys = django_settings_overrides.keys()
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply openedx/features/edly/utils.py:325]: done

File: openedx/features/edly/utils.py [hunk 3]
```diff
@@ -313,3 +314,35 @@ def filter_courses_based_on_org(request, all_courses):
     filtered_courses = [course for course in list(all_courses) if course.org == edx_org]
 
     return filtered_courses
+
+def clean_django_settings_override(django_settings_overrides):
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply openedx/features/edly/utils.py:318]: done

File: openedx/core/djangoapps/site_configuration/models.py [hunk 4]
```diff
@@ -60,6 +62,15 @@ def get_value(self, name, default=None):
...
 
+    def clean(self):
+        """
+        Add check for allowed django settings override.
+        """
+        super(SiteConfiguration, self).clean()
+        if self.enabled:
+            clean_django_settings_override(self.get_value('DJANGO_SETTINGS_OVERRIDE', None))
+
+
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply openedx/core/djangoapps/site_configuration/models.py:73]: done

File: cms/envs/common.py [hunk 5]
```diff
@@ -590,6 +590,12 @@
 EDLY_COOKIE_SECRET_KEY = 'EDLY-COOKIE-SECRET-KEY'
 EDLY_JWT_ALGORITHM = 'HS256'
 
+ALLOWED_DJANGO_SETTINGS_OVERRIDE = [
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply cms/envs/common.py:593]: done

File: openedx/features/edly/utils.py [hunk 6]
```diff
@@ -313,3 +314,35 @@ def filter_courses_based_on_org(request, all_courses):
     filtered_courses = [course for course in list(all_courses) if course.org == edx_org]
 
     return filtered_courses
+
+def clean_django_settings_override(django_settings_override):
```
Comment (zubair-arbi) [inline openedx/features/edly/utils.py:318]: Modify this method like following:
```python
def clean_django_settings_override(django_settings_override):
    """
    Enforce only allowed django settings to be overridden.
    """
    if not django_settings_override:
        return

    django_settings_override_keys = django_settings_override.keys()
    disallowed_override_keys = list(set(django_settings_override_keys) - set(settings.ALLOWED_DJANGO_SETTINGS_OVERRIDE))
    updated_override_keys = list(set(django_settings_override_keys) - set(disallowed_override_keys))
    missing_override_keys = list(set(settings.ALLOWED_DJANGO_SETTINGS_OVERRIDE) - set(updated_override_keys))
    
    validation_errors = []
    if disallowed_override_keys:
        disallowed_override_keys_string = ', '.join(disallowed_override_keys)
        validation_errors.append(
            ValidationError(
                _('Django settings override(s) "%(disallowed_overrides)s" is/are not allowed to be overridden.'),
                params={'disallowed_overrides': disallowed_override_keys_string},
            )
        )
        
    if missing_override_keys:
        missing_override_keys_string = ', '.join(missing_override_keys)
        validation_errors.append(
            ValidationError(
                _('Django settings override(s) "%(missing_overrides)s" is/are missing.'),
                params={'missing_overrides': missing_override_keys_string},
            )
        )

    if validation_errors:
        raise ValidationError(validation_errors)
```

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
Added one more feedback.

File: openedx/features/edly/utils.py (hunk 6 above)
Comment (rehan99000) [reply openedx/features/edly/utils.py:318]: done

File: openedx/features/edly/utils.py [hunk 7]
```diff
@@ -313,3 +314,37 @@ def filter_courses_based_on_org(request, all_courses):
...
+    updated_override_keys = list(set(django_settings_override_keys) - set(disallowed_override_keys))
+    missing_override_keys = list(set(settings.ALLOWED_DJANGO_SETTINGS_OVERRIDE) - set(updated_override_keys))
+
+    validation_errors = []
+    if disallowed_override_keys:
+        disallowed_override_keys_string = ', '.join(disallowed_override_keys)
+        validation_errors.append(
+            ValidationError(
+                _('Django settings override(s) "%(disallowed_overrides)s" is/are not allowed to be overridden.'),
+                params={'disallowed_overrides': disallowed_override_keys_string},
```
Comment (zubair-arbi) [inline openedx/features/edly/utils.py:336]: `disallowed_overrides` -> `disallowed_override_keys`

**Review by zubair-arbi — APPROVED** [review]
Changes looks good 👍 

Please check the client onboarding document one more time and @rehan99000 rebase your branch.

**Review by asadmanzoor93 — APPROVED** [review]

**Pull request merged**